File: src/DashSpec.Core/Parsing/CardVisibilityParser.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal static class CardVisibilityParser
{
    public static CardVisibilityRule ParseFilterWhen(TokenReader reader, string cardId, string filterName)
    {
        if (string.IsNullOrWhiteSpace(filterName))
        {
            throw new DashSpecParseException($"Card '{cardId}': when requires filter name or oversize.");
        }

        if (reader.TryKeyword("empty"))
        {
            return new CardVisibilityRule(filterName, CardVisibilityMode.WhenEmpty);
        }

        if (reader.TryKeyword("set"))
        {
            reader.SkipNewlines();
            return new CardVisibilityRule(
                filterName,
                CardVisibilityMode.WhenSet,
                TryReadMessageBody(reader, cardId, filterName));
        }

        reader.SkipNewlines();
        if (!reader.IsEof && reader.TryKeyword("message"))
        {
            reader.Expect(TokenKind.Eq);
            var inlineMessage = reader.ReadString();
            reader.SkipNewlines();
            ExpectWhenBlockEnd(reader, filterName);
            return new CardVisibilityRule(filterName, CardVisibilityMode.WhenSet, inlineMessage);
        }

        if (!reader.IsEof && !IsWhenBlockEnd(reader, filterName))
        {
            return new CardVisibilityRule(
                filterName,
                CardVisibilityMode.WhenSet,
                ReadMessageBody(reader, cardId, filterName));
        }

        return new CardVisibilityRule(filterName, CardVisibilityMode.WhenSet);
    }

    public static string ParseOversizeWhen(TokenReader reader, string cardId)
    {
        reader.SkipNewlines();
        return ReadMessageBody(reader, cardId, "oversize");
    }

    private static string? TryReadMessageBody(TokenReader reader, string cardId, string filterName) =>
        reader.IsEof || IsWhenBlockEnd(reader, filterName)
            ? null
            : ReadMessageBody(reader, cardId, filterName);

    private static string ReadMessageBody(TokenReader reader, string cardId, string filterName)
    {
        reader.SkipNewlines();
        string? message = null;

        while (!IsWhenBlockEnd(reader, filterName) && !reader.IsEof)
        {
            reader.SkipNewlines();
            if (IsWhenBlockEnd(reader, filterName))
            {
                break;
            }

            var key = reader.ReadIdent();
            if (!string.Equals(key, "message", StringComparison.OrdinalIgnoreCase))
            {
                throw new DashSpecParseException(
                    $"Card '{cardId}': when block supports only message = \"…\".");
            }

            reader.Expect(TokenKind.Eq);
            message = reader.ReadString();
            reader.SkipNewlines();
        }

        ExpectWhenBlockEnd(reader, filterName);

        if (string.IsNullOrWhiteSpace(message))
        {
            throw new DashSpecParseException($"Card '{cardId}': when message cannot be empty.");
        }

        return message;
    }

    private static bool IsWhenBlockEnd(TokenReader reader, string? filterName = null) =>
        BlockSyntax.IsBlockEnd(reader, "when") ||
        BlockSyntax.IsBlockEnd(reader, "oversize") ||
        (filterName is not null && BlockSyntax.IsBlockEnd(reader, filterName));

    private static void ExpectWhenBlockEnd(TokenReader reader, string? filterName = null)
    {
        if (filterName is not null && BlockSyntax.IsBlockEnd(reader, filterName))
        {
            BlockSyntax.ExpectBlockEnd(reader, filterName);
            return;
        }

        if (BlockSyntax.IsBlockEnd(reader, "oversize"))
        {
            BlockSyntax.ExpectBlockEnd(reader, "oversize");
            return;
        }

        BlockSyntax.ExpectBlockEnd(reader, "when");
    }
}

File: src/DashSpec.Modeling.Parse/Card/CardVisibilityParser.fs
namespace DashSpec.Modeling.Parse.Card

open System
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse
open DashSpec.Modeling.Parse.Lexing

module CardVisibilityParser =

    let private isWhenBlockEnd (reader: TokenReader) (filterName: string option) =
        BlockSyntax.isBlockEnd reader "when" None
        || BlockSyntax.isBlockEnd reader "oversize" None
        || (filterName.IsSome && BlockSyntax.isBlockEnd reader filterName.Value None)

    let private expectWhenBlockEnd (reader: TokenReader) (filterName: string option) =
        if filterName.IsSome && BlockSyntax.isBlockEnd reader filterName.Value None then
            BlockSyntax.expectBlockEnd reader filterName.Value None
        elif BlockSyntax.isBlockEnd reader "oversize" None then
            BlockSyntax.expectBlockEnd reader "oversize" None
        else
            BlockSyntax.expectBlockEnd reader "when" None

    let private readMessageBody (reader: TokenReader) (cardId: string) (filterName: string) =
        reader.SkipNewlines()
        let mutable message: string option = None

        while not (isWhenBlockEnd reader (Some filterName)) && not reader.IsEof do
            reader.SkipNewlines()
            if isWhenBlockEnd reader (Some filterName) then ()
            else
                let key = reader.ReadIdent()
                if not (String.Equals(key, "message", StringComparison.OrdinalIgnoreCase)) then
                    raise (DashSpecParseException($"Card '{cardId}': when block supports only message = \"…\"."))
                reader.Expect TokenKind.Eq
                message <- Some(reader.ReadString())
                reader.SkipNewlines()

        expectWhenBlockEnd reader (Some filterName)

        match message with
        | None | Some (null | "") ->
            raise (DashSpecParseException($"Card '{cardId}': when message cannot be empty."))
        | Some value -> value

    let private tryReadMessageBody (reader: TokenReader) (cardId: string) (filterName: string) =
        if reader.IsEof || isWhenBlockEnd reader (Some filterName) then None
        else Some(readMessageBody reader cardId filterName)

    let parseFilterWhen (reader: TokenReader) (cardId: string) (filterName: string) =
        if String.IsNullOrWhiteSpace filterName then
            raise (DashSpecParseException($"Card '{cardId}': when requires filter name or oversize."))

        if reader.TryKeyword "empty" then
            { FilterName = filterName; Mode = CardVisibilityMode.WhenEmpty; Message = None }
        elif reader.TryKeyword "set" then
            reader.SkipNewlines()
            { FilterName = filterName
              Mode = CardVisibilityMode.WhenSet
              Message = tryReadMessageBody reader cardId filterName }
        else
            reader.SkipNewlines()
            if not reader.IsEof && reader.TryKeyword "message" then
                reader.Expect TokenKind.Eq
                let inlineMessage = reader.ReadString()
                reader.SkipNewlines()
                expectWhenBlockEnd reader (Some filterName)
                { FilterName = filterName; Mode = CardVisibilityMode.WhenSet; Message = Some inlineMessage }
            elif not reader.IsEof && not (isWhenBlockEnd reader (Some filterName)) then
                { FilterName = filterName
                  Mode = CardVisibilityMode.WhenSet
                  Message = Some(readMessageBody reader cardId filterName) }
            else
                { FilterName = filterName; Mode = CardVisibilityMode.WhenSet; Message = None }

    let parseOversizeWhen (reader: TokenReader) (cardId: string) =
        reader.SkipNewlines()
        readMessageBody reader cardId "oversize"
